// redux-orm/lib/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ID_ARG_KEY_SELECTOR = exports.ALL_INSTANCES = exports.STATE_FLAG = exports.FAILURE = exports.SUCCESS = exports.ORDER_BY = exports.EXCLUDE = exports.FILTER = exports.CREATE = exports.DELETE = exports.UPDATE = void 0;
var UPDATE = "REDUX_ORM_UPDATE";
exports.UPDATE = UPDATE;
var DELETE = "REDUX_ORM_DELETE";
exports.DELETE = DELETE;
var CREATE = "REDUX_ORM_CREATE";
exports.CREATE = CREATE;
var FILTER = "REDUX_ORM_FILTER";
exports.FILTER = FILTER;
var EXCLUDE = "REDUX_ORM_EXCLUDE";
exports.EXCLUDE = EXCLUDE;
var ORDER_BY = "REDUX_ORM_ORDER_BY";
exports.ORDER_BY = ORDER_BY;
var SUCCESS = "SUCCESS";
exports.SUCCESS = SUCCESS;
var FAILURE = "FAILURE"; // for detecting ORM state objects

exports.FAILURE = FAILURE;
var STATE_FLAG = "@@_______REDUX_ORM_STATE_FLAG"; // for caching selectors based on their ID argument

exports.STATE_FLAG = STATE_FLAG;
var ALL_INSTANCES = Symbol("REDUX_ORM_ALL_INSTANCES");
exports.ALL_INSTANCES = ALL_INSTANCES;

var ID_ARG_KEY_SELECTOR = function ID_ARG_KEY_SELECTOR(_state, idArg) {
  return typeof idArg === "undefined" ? ALL_INSTANCES : idArg;
};

exports.ID_ARG_KEY_SELECTOR = ID_ARG_KEY_SELECTOR;
