// redux-orm/lib/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "QuerySet", {
  enumerable: true,
  get: function get() {
    return _QuerySet["default"];
  }
});
Object.defineProperty(exports, "Model", {
  enumerable: true,
  get: function get() {
    return _Model["default"];
  }
});
Object.defineProperty(exports, "ORM", {
  enumerable: true,
  get: function get() {
    return _ORM.ORM;
  }
});
Object.defineProperty(exports, "Session", {
  enumerable: true,
  get: function get() {
    return _Session["default"];
  }
});
Object.defineProperty(exports, "createReducer", {
  enumerable: true,
  get: function get() {
    return _redux.createReducer;
  }
});
Object.defineProperty(exports, "createSelector", {
  enumerable: true,
  get: function get() {
    return _redux.createSelector;
  }
});
Object.defineProperty(exports, "ForeignKey", {
  enumerable: true,
  get: function get() {
    return _ForeignKey["default"];
  }
});
Object.defineProperty(exports, "ManyToMany", {
  enumerable: true,
  get: function get() {
    return _ManyToMany["default"];
  }
});
Object.defineProperty(exports, "OneToOne", {
  enumerable: true,
  get: function get() {
    return _OneToOne["default"];
  }
});
Object.defineProperty(exports, "Attribute", {
  enumerable: true,
  get: function get() {
    return _Attribute["default"];
  }
});
Object.defineProperty(exports, "fk", {
  enumerable: true,
  get: function get() {
    return _fields.fk;
  }
});
Object.defineProperty(exports, "many", {
  enumerable: true,
  get: function get() {
    return _fields.many;
  }
});
Object.defineProperty(exports, "oneToOne", {
  enumerable: true,
  get: function get() {
    return _fields.oneToOne;
  }
});
Object.defineProperty(exports, "attr", {
  enumerable: true,
  get: function get() {
    return _fields.attr;
  }
});
exports["default"] = exports.Backend = exports.Schema = void 0;

var _QuerySet = _interopRequireDefault(require("./QuerySet"));

var _Model = _interopRequireDefault(require("./Model"));

var _ORM = require("./ORM");

var _Session = _interopRequireDefault(require("./Session"));

var _redux = require("./redux");

var _ForeignKey = _interopRequireDefault(require("./fields/ForeignKey"));

var _ManyToMany = _interopRequireDefault(require("./fields/ManyToMany"));

var _OneToOne = _interopRequireDefault(require("./fields/OneToOne"));

var _Attribute = _interopRequireDefault(require("./fields/Attribute"));

var _fields = require("./fields");

var Schema = _ORM.DeprecatedSchema;
exports.Schema = Schema;

var Backend = function RemovedBackend() {
  throw new Error("Having a custom Backend instance is now unsupported. " + "Documentation for database customization is upcoming, for now " + "please look at the db folder in the source.");
};

exports.Backend = Backend;
var _default = _Model["default"];
exports["default"] = _default;
